panel/src/helpers/upload.js

```diff
@@ -96,4 +98,46 @@ export default async (file, params) => {
 
 		xhr.send(data);
 	});
-};
+}
+
+/**
+ * Uploads a file in chunks
+ * @param {File} file - file to upload
+ * @param {Object} params - upload options (see `upload` method for details)
+ * @param {number} size - chunk size in bytes (default: 5 MB)
+ */
+export async function uploadAsChunks(file, params, size = 5242880) {
+	const parts = Math.ceil(file.size / size);
+	const id = random(4).toLowerCase();
+	let response;
+
+	for (let i = 0; i < parts; i++) {
+		const start = i * size;
+		const end = Math.min(start + size, file.size);
+		const chunk = parts > 1 ? file.slice(start, end, file.type) : file;
+
+		// when more than one part, add flag to
+		// recognize chunked upload and its last chunk
+		if (parts > 1) {
+			params.headers = {
+				...params.headers,
+				"Upload-Length": file.size,
+				"Upload-Offset": start,
+				"Upload-Id": id
+			};
+		}
+
+		response = await upload(chunk, {
+			...params,
+			// calculate the total progress based on chunk progress
+			progress: (xhr, chunk, percent) => {
+				const total = (i + percent / 100) / parts;
```

Comment:
This calculation causes the progress to jump quickly at the end if the last chunk is smaller than the rest. E.g. file is 91 MB and chunk size is 10 MB. Then chunk 10/10 is only 1 MB, so the last 10% of progress pass 10 times faster.

Maybe we can implement it as transmitted size (start offset + current chunk size * percent) compared to the total file size.

Reply:
@lukasbestle 

```suggestion
				const offset = i * size;
				const current = size * (percent / 100);
				const total = (offset + current) / file.size;
```

?

Reply:
I think instead of `size` (which again does not respect that the last chunk may be smaller) it should be `chunk.size`. And instead of `offset` I think it should be `start`. Rest looks great.